### ISSUE TITLE
Build fails with GCC 7

### ISSUE THREAD

**Issue by mmuehlenhoff** [facebook/hhvm#7961]
I tried to compile HHVM 3.21 with GCC 7, but it fails thirdparty/mcrouter. There's a pre-existing issue for mcrouter at https://github.com/facebook/mcrouter/issues/200 (with the fix being that the affected code was removed entirely), could this also be pulled into 3.21.x? GCC 7 is now the default in Debian buster and it would be great to be able to build the latest LTS with the default compiler.

**Comment by mmuehlenhoff** [comment]
This is caused by an header cleanup in the STL in GCC 7, see the section "Header dependency changes" from https://gcc.gnu.org/gcc-7/porting_to.html

Please consider applying the following patch (and ideally merge to 3.21.x, so that future LTS point releases can be build with GCC 7). I've verified that it fixes the build with GCC 7.1:

```
diff --git a/third-party/mcrouter/src/mcrouter/lib/cycles/Cycles.h b/third-party/mcrouter/src/mcrouter/lib/cycles/Cycles.h
index 4f57e1290..224e92ca7 100644
--- a/third-party/mcrouter/src/mcrouter/lib/cycles/Cycles.h
+++ b/third-party/mcrouter/src/mcrouter/lib/cycles/Cycles.h
@@ -9,6 +9,7 @@
  */
 #pragma once
 
+#include <functional>
 #include <cstdint>
 #include <memory>
 #include <string>
```

**Comment by fredemmott** [comment]
Sorry for the delay; we won't be backporting a full mcrouter update to 3.21 - too high a risk of breaking other things. If you send a PR to https://github.com/facebook/hhvm/tree/master/patches against the 3.21 branch and sign the CLA if you haven't already, I'll be happy to merge it.

I'm not able to take that patch unless it's contributed as a PR under the CLA, or to reproduce it, and we don't currently target GCC7.
